src/app/decorations/qgsdecorationscalebar.cpp

```diff
@@ -243,34 +241,20 @@ void QgsDecorationScaleBar::render( const QgsMapSettings &mapSettings, QgsRender
   const float deviceHeight = static_cast<float>( device->height() ) / context.devicePixelRatio();
   const float deviceWidth = static_cast<float>( device->width() ) / context.devicePixelRatio();
   const Qgis::DistanceUnit preferredUnits = QgsProject::instance()->distanceUnits();
-  Qgis::DistanceUnit scaleBarUnits = mapSettings.mapUnits();
 
   //Get map units per pixel
   const double scaleBarUnitsPerPixel = ( mapWidth( mapSettings ) / mapSettings.outputSize().width() ) * QgsUnitTypes::fromUnitToUnitFactor( mSettings.units(), preferredUnits );
-  scaleBarUnits = preferredUnits;
+  Qgis::DistanceUnit scaleBarUnits = preferredUnits;
 
   // Exit if the canvas width is 0 or layercount is 0 or QGIS will freeze
   if ( mapSettings.layers().isEmpty() || !deviceWidth || !scaleBarUnitsPerPixel )
     return;
 
-  double unitsPerSegment = mPreferredSize;
+  //Calculate size of scale bar for 25% of map window
+  double scaleBarWidth = deviceWidth * 0.25;
 
-  //Calculate size of scale bar for preferred number of map units
-  double scaleBarWidth = mPreferredSize / scaleBarUnitsPerPixel;
-
-  //If scale bar is very small reset to 1/4 of the canvas wide
-  if ( scaleBarWidth < 30 )
```

Comment:
Isn't the current logic:

- start with the preferred size
- if it's less than 30 **pixels**, resize to one quarter of the window
- if it's more than 30% of the window width, resize down

So the preferred size WILL be used if it's > 30 pix and < 30% of window size

Reply:
Thank you very much for the comment.
The problem is that the inserted value in the dialogue depends on the default range in the QSpinBox class with a maximum of 99. I do not know if it is possible to change that maximum default. So the preferred size will only be recognized if the size of the map canvas and the current map scale suits the inserted value. In my tries that is hardly the case and at least larger numbers are needed. 
I think, the user gets more confused if there is the chance to enter a value which is not working properly in most cases. So maybe it is better not to offer the option at all? 
Maybe one alternative is to insert a value as a percentage of the map canvas instead of a value in map units. Otherwise the default maximum has to be changed.
I think most people are using the "Automatically snap to round number on resize" check box to get some readable numbers at the scale bar so a approximately fixed length will be fine.
It is my first commit to qgis done at the contributor meeting in Bratislava. I discussed the possible options there with some experienced people and we came to the decision to size it to 25% of the map canvas. In the end we are all do not really use these feature (scale bar decoration) and it is still the question from the original issue #15647 how it _should_ work.
I am thankful for any suggestion to fix this finally!